### PR TITLE
Major Cleanup #1

### PR DESCRIPTION
**Pull request by Mehrad0711**
A cleanup for genienlp has been long overdue!

We've collected quite some technical debt which we need to address. We have created a [project dashboard](https://github.com/stanford-oval/genienlp/projects/4) to keep track of issues we want to address. I've moved the cards this PR addresses to "In progress".

The goal of this series of cleanups is to simplify the code, make it more readable, easier to modify, and more inviting for outside collaborators.

There are no functional changes (besides dropping almond_multilingual tasks and associated code) - tests are still passing.

We're dropping support for old generic datasets and associated metrics which were inherited from decaNLP code.
Those implementations are obsolete and should be replaced with what is now accessible through [datasets library](https://github.com/huggingface/datasets). Please see [issue](https://github.com/stanford-oval/genienlp/issues/231). If you want support for a new dataset, feel free to submit a PR for it.

### CODE REVIEW

File: genienlp/metrics.py
Comment (s-jse) [inline genienlp/metrics.py:404]: Please bring back Rouge, perhaps using HF dataset. It is not just for summarization, it is also used in many natural_seq2seq tasks like paraphrasing. I know it is not on by default, but we do use it from time to time.

Comment (s-jse) [inline genienlp/metrics.py:404]: Please bring back all Rouge scores (1, 2, L), perhaps using HF dataset. It is not just for the -now deleted- summarization task, it is also used in many `natural_seq2seq` tasks like paraphrasing. I know it is not used by default, but we do use it from time to time.

File: genienlp/metrics.py
Comment (s-jse) [inline genienlp/metrics.py]: Address the TODO, list what metrics we support.

File: genienlp/models/base.py
Comment (s-jse) [inline genienlp/models/base.py:664]: This hierarchy of validation/generation functions, their naming and the naming of their outputs are not quite right. For example, when validating a `TransformerSeq2Seq` model, we have the following:

we call `validate()` in `validate.py`
which calls `generate_with_model()`
which calls `validate()` in `models/base.py` which outputs a `GenerationOutput` object
which calls `generate()` in `models/transformer_seq2seq.py`
which calls `generate()` from HF's `transformers` package.

First, there are too many levels, and second, it is really confusing.
You should define how many levels are needed, and move around the functionality until each level does exactly one clearly defined thing. We can discuss this more.

File: genienlp/models/base.py
Comment (s-jse) [inline genienlp/models/base.py:690]: You have put `TransformerForSequenceClassification` and `TransformerForTokenClassification` under the same parent class.
This comment made me wonder how is it that their validation can have the same implementation and therefore be in the parent class. Don't these models outputs at the very least have different dimensions? One is an `n*c`, the other one is `1*c` (n is sequence length, c is the number of classes)

File: genienlp/predict.py
Comment (s-jse) [inline genienlp/predict.py:432]: So now `pred_src_languages` and `pred_tgt_languages` are only used for determining the language token of pre-trained models that support it?

**Review by s-jse — CHANGES_REQUESTED** [review]
Thanks, overall is a good cleanup, but you should make a few changes before merging it. Please see the individual comments.
Also, have you tested this code with old non-multilingual models? What about the genie-k8s scripts? Especially since you have removed training/prediction arguments.

File: genienlp/predict.py
Comment (Mehrad0711) [reply genienlp/predict.py:432]: There are no functional changes. Before you could pass a string of languages (joined by '+') for each task. It was a convenient way of evaluating on multiple languages in one run (e.g. for almond_multilingual) tasks. 
I removed that option since almond_multilingual is gone and the same task can be achieved by running predict multiple times once for each language.

File: genienlp/models/base.py
Comment (Mehrad0711) [reply genienlp/models/base.py:690]: This exact code does the unification. It turns 2D (batch, class) outputs of sequence classification into 3D (batch, length, class) to conform to 3D output of token classification. 
We then turn each output into a string in the next block of code so everything is flattened out.

File: genienlp/metrics.py
Comment (Mehrad0711) [reply genienlp/metrics.py]: Improving documentation is not of my goals in this PR. Sorry...
I'd rather leave this to future PRs or someone else to address.

File: genienlp/models/base.py
Comment (Mehrad0711) [reply genienlp/models/base.py:664]: I agree. Lemme see what I can do.

File: genienlp/train.py
Comment (s-jse) [inline genienlp/train.py]: The new changes to the validation code have definitely made it cleaner. Can you change the naming of `validate()` and `do_validate()` in train.py to `validate_during_training()` and `do_validate_during_training()` to further differentiate them from `validate()` defined elsewhere?

File: genienlp/train.py
Comment (s-jse) [inline genienlp/train.py]: Can you also rename `generation_output` variable (here and everywhere else) and the `GenerationOutput` class to `validation_output` and `ValidationOutput`? This way, the term "generation" (generating text output) and "validation" (generating text output and then using gold answers to potentially calculate confidence scores or metrics) will have the same meaning everywhere.

File: genienlp/train.py
Comment (Mehrad0711) [reply genienlp/train.py]: The only `validate()` function (not attached to a class) we have now is in "train.py". Not sure if we should namespace it. If you're worried about confusion with `validate()` methods in "base.py" I can rename it to `evaluate()` and `do_evaluate()` as that's a shorter and more conventional name (HF use it also). What do you think?

File: genienlp/train.py
Comment (Mehrad0711) [reply genienlp/train.py]: Sure

File: genienlp/train.py
Comment (s-jse) [reply genienlp/train.py]: Validate and evaluate are often used interchangeably, so doesn't help with confusion. I just think validate_during_training() is more descriptive, and e.g. made me quickly realize I should not use it for the new prediction code I plan to write.

Comment (Mehrad0711) [reply genienlp/train.py]: Ok let's go with your suggestion

**Review by s-jse — APPROVED** [review]

**Pull request merged**
